File: sdk/python/tests/e2e_edge.py
"""E2E: edge cases — payload limits, unicode, pipelining, bulk frames,
isolation, empty long-poll, stable pagination, crash + restart reconnect,
idempotent custom ids, per-job results under concurrency, option extremes.

Mirrors the TypeScript SDK's e2e-edge.ts so both SDKs prove the same wire
behavior at the boundaries.
"""

from __future__ import annotations

import threading
import time

from harness import Server, test, unique_name, wait_until

from bunqueue import Queue, Worker


@test
def edge_2mb_payload_roundtrip(server: Server) -> None:
    with Queue(unique_name("big"), port=server.port) as queue:
        blob = "x" * (2 * 1024 * 1024)
        job = queue.add("blob", {"blob": blob})
        fetched = queue.get_job(job.id)
        assert fetched is not None and fetched.data.get("blob") == blob, "2MB payload must survive"


@test
def edge_unicode_nested_payload(server: Server) -> None:
    with Queue(unique_name("uni"), port=server.port) as queue:
        payload = {
            "emoji": "🚀🔥💯",
            "cjk": "以呂波耳本部止",
            "rtl": "مرحبا بالعالم",
            "nested": {"list": [1, "två", {"deep": "ключ"}], "null": None, "bool": True},
        }
        job = queue.add("uni", payload)
        fetched = queue.get_job(job.id)
        assert fetched is not None
        for key in ("emoji", "cjk", "rtl"):
            assert fetched.data.get(key) == payload[key], f"{key} corrupted by msgpack roundtrip"
        assert fetched.data.get("nested", {}).get("list", [None])[-1] == {"deep": "ключ"}


@test
def edge_200_concurrent_adds_pipeline(server: Server) -> None:
    """200 adds from 8 threads pipeline on one socket without cross-talk."""
    with Queue(unique_name("pipe"), port=server.port) as queue:
        ids: list = []
        lock = threading.Lock()

        def producer(base: int) -> None:
            for i in range(25):
                job = queue.add("evt", {"n": base + i})
                with lock:
                    ids.append(job.id)

        threads = [threading.Thread(target=producer, args=(t * 25,)) for t in range(8)]
        for t in threads:
            t.start()
        for t in threads:
            t.join()
        assert len(ids) == 200 and len(set(ids)) == 200, "every add must get a unique id"
        assert wait_until(lambda: queue.count() == 200, 15)


@test
def edge_add_bulk_1000_one_frame(server: Server) -> None:
    with Queue(unique_name("bulk"), port=server.port) as queue:
        ids = queue.add_bulk([{"name": "evt", "data": {"i": i}} for i in range(1000)])
        assert len(ids) == 1000 and len(set(ids)) == 1000
        assert wait_until(lambda: queue.count() == 1000, 15)


@test
def edge_multi_queue_isolation(server: Server) -> None:
    with Queue(unique_name("iso-a"), port=server.port) as qa:
        with Queue(unique_name("iso-b"), port=server.port) as qb:
            qa.add("a", {"q": "a"})
            qb.add("b", {"q": "b"})
            assert wait_until(lambda: qa.count() == 1 and qb.count() == 1, 10)
            done: list = []
            worker = Worker(
                qa.name, lambda j: done.append(j.data["q"]) or "ok", port=server.port, poll_timeout_ms=300
            )
            try:
                worker.start()
                assert wait_until(lambda: len(done) == 1, 15)
                time.sleep(0.5)
                assert done == ["a"], f"worker on queue A must never see queue B: {done}"
                assert qb.count() == 1, "queue B untouched"
            finally:
                worker.close(timeout=10)


@test
def edge_empty_long_poll_returns_no_jobs(server: Server) -> None:
    with Queue(unique_name("empty"), port=server.port) as queue:
        started = time.monotonic()
        response = queue.connection.call(
            {"cmd": "PULLB", "queue": queue.name, "count": 1, "timeout": 800, "owner": "edge-test"}
        )
        elapsed = time.monotonic() - started
        assert not response.get("jobs"), "empty queue must yield no jobs"
        assert elapsed >= 0.5, f"long-poll must actually wait, returned in {elapsed:.2f}s"


@test
def edge_get_jobs_pagination_stable(server: Server) -> None:
    with Queue(unique_name("page"), port=server.port) as queue:
        queue.add_bulk([{"name": "evt", "data": {"i": i}} for i in range(30)])
        assert wait_until(lambda: len(queue.get_jobs(["waiting"], 0, 30)) == 30, 15)
        page1 = queue.get_jobs(["waiting"], 0, 10)
        page2 = queue.get_jobs(["waiting"], 10, 20)
        page3 = queue.get_jobs(["waiting"], 20, 30)
        ids = [j.id for j in page1 + page2 + page3]
        assert len(ids) == 30 and len(set(ids)) == 30, "pages must not overlap or skip"


@test
def edge_worker_survives_server_crash_restart(server: Server) -> None:
    """Kill the server mid-run, restart on the same port: producer and worker
    reconnect on their own and drain the remaining jobs (zero manual steps)."""
    own = Server()
    own.start()
    try:
        queue = Queue(unique_name("crash"), port=own.port)
        done: set = set()
        worker = Worker(queue.name, lambda j: done.add(j.data["i"]) or "ok", port=own.port, poll_timeout_ms=300)
        worker.on("error", lambda e: None)  # reconnect noise expected
        worker.start()
        for i in range(5):
            queue.add("evt", {"i": i}, attempts=3)
        assert wait_until(lambda: len(done) == 5, 15)

        assert own.proc is not None
        own.proc.kill()
        own.proc.wait(timeout=10)
        own.proc = None
        time.sleep(0.5)
        own.start()  # same port, same data dir

        for i in range(5, 10):
            queue.add("evt", {"i": i}, attempts=3)
        assert wait_until(lambda: len(done) == 10, 30), f"drained only {sorted(done)}"
        worker.close(timeout=10)
        queue.close()
    finally:
        own.stop()


@test
def edge_duplicate_custom_ids_idempotent(server: Server) -> None:
    with Queue(unique_name("dup"), port=server.port) as queue:
        queue.add_bulk(
            [
                {"name": "evt", "data": {"v": 1}, "job_id": "edge-dup-1"},
                {"name": "evt", "data": {"v": 2}, "job_id": "edge-dup-1"},
                {"name": "evt", "data": {"v": 3}, "job_id": "edge-dup-2"},
            ]
        )
        assert wait_until(lambda: queue.count() == 2, 15), "duplicate custom id must dedupe"
        first = queue.get_job_by_custom_id("edge-dup-1")
        assert first is not None and first.data.get("v") == 1, "first write wins on dedupe"


@test
def edge_results_preserved_per_job_under_concurrency(server: Server) -> None:
    with Queue(unique_name("res"), port=server.port) as queue:
        worker = Worker(
            queue.name, lambda j: {"double": j.data["i"] * 2}, port=server.port, concurrency=8, poll_timeout_ms=300
        )
        try:
            worker.start()
            jobs = [queue.add("calc", {"i": i}) for i in range(20)]
            assert wait_until(
                lambda: queue.get_job_counts().get("completed", 0) >= 20, 20
            )
            for i, job in enumerate(jobs):
                assert queue.get_result(job.id) == {"double": i * 2}, f"result mixed up for job {i}"
        finally:
            worker.close(timeout=10)


@test
def edge_batch_size_zero_clamps_to_one(server: Server) -> None:
    with Queue(unique_name("bz"), port=server.port) as queue:
        done: list = []
        worker = Worker(
            queue.name, lambda j: done.append(j.id) or "ok", port=server.port, batch_size=0, poll_timeout_ms=300
        )
        try:
            assert worker.batch_size == 1, f"batch_size 0 must clamp to 1, got {worker.batch_size}"
            worker.start()
            queue.add("t", {"x": 1})
            assert wait_until(lambda: len(done) >= 1, 15)
        finally:
            worker.close(timeout=10)


@test
def edge_per_job_stack_trace_limit_honored(server: Server) -> None:
    """A job asking for a deeper stack (stack_trace_limit=30) must get more
    than the 10-line default AND keep the raise site."""
    with Queue(unique_name("stk30"), port=server.port) as queue:

        def deep(n: int) -> int:
            if n <= 0:
                raise RuntimeError("BOOM-DEEP-30")
            return deep(n - 1)

        worker = Worker(queue.name, lambda j: deep(25), port=server.port, poll_timeout_ms=300)
        try:
            worker.start()
            job = queue.add("t", {}, attempts=1, stack_trace_limit=30)
            assert wait_until(lambda: queue.get_state(job.id) == "failed", 20)
            failed = queue.get_job(job.id)
            assert failed is not None
            stack = failed.stacktrace or []
            assert len(stack) > 10, f"per-job stack_trace_limit ignored: {len(stack)} lines"
            assert any("BOOM-DEEP-30" in line for line in stack), "raise site truncated away"
        finally:
            worker.close(timeout=10)

File: sdk/typescript/tests/e2e-edge.ts
/** E2E: edge cases — payload limits, unicode, pipelining, reconnect, isolation. */

import { type ChildProcess, spawn } from 'node:child_process';
import { mkdtempSync } from 'node:fs';
import { connect } from 'node:net';
import { tmpdir } from 'node:os';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { Queue } from '../dist/legacy.js';
import {
  assert,
  assertEq,
  getPort,
  makeQueue,
  makeWorker,
  qname,
  sleep,
  test,
  waitFor,
} from './harness.ts';

const REPO_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..', '..', '..');

/** Spawn a bunqueue server on a FIXED port and wait until it accepts TCP. */
async function spawnFixedServer(port: number): Promise<ChildProcess> {
  const dataDir = mkdtempSync(join(tmpdir(), 'bunqueue-edge-'));
  const proc = spawn('bun', ['src/main.ts'], {
    cwd: REPO_ROOT,
    env: {
      ...process.env,
      TCP_PORT: String(port),
      HTTP_PORT: '0',
      BUNQUEUE_DATA_PATH: join(dataDir, 'bunq.db'),
    },
    stdio: 'ignore',
  });
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const up = await new Promise<boolean>((res) => {
      const s = connect({ host: '127.0.0.1', port }, () => {
        s.destroy();
        res(true);
      });
      s.on('error', () => res(false));
      s.setTimeout(500, () => {
        s.destroy();
        res(false);
      });
    });
    if (up) return proc;
    await sleep(100);
  }
  proc.kill();
  throw new Error('fixed-port bunqueue server did not start within 15s');
}

test('edge: 2MB payload roundtrip', async () => {
  const queue = makeQueue<{ blob: string }>('edge-big');
  const blob = 'x'.repeat(2 * 1024 * 1024);
  const job = await queue.add('big', { blob });
  const fetched = await queue.getJob(job.id);
  assert(fetched !== null, 'job found');
  assertEq((fetched!.data as { blob: string }).blob.length, blob.length, 'payload intact');
  queue.close();
});

test('edge: unicode + nested payload survives msgpack', async () => {
  const queue = makeQueue('edge-unicode');
  const data = {
    emoji: '🚀🔥 “quotes” — dash',
    jp: 'こんにちは世界',
    nested: { list: [1, 2.5, null, true, 'x'], deep: { a: { b: { c: 'd' } } } },
    negative: -42,
    float: 12.34567,
  };
  const job = await queue.add('uni', data);
  const fetched = await queue.getJob(job.id);
  const got = fetched?.data as typeof data & { name: string };
  assertEq(got.emoji, data.emoji, 'emoji intact');
  assertEq(got.jp, data.jp, 'japanese intact');
  assertEq(got.nested.deep.a.b.c, 'd', 'deep nesting intact');
  assertEq(got.negative, -42, 'negative int intact');
  assertEq(got.float, 12.34567, 'float intact');
  queue.close();
});

test('edge: 200 concurrent adds pipeline on one socket', async () => {
  const queue = makeQueue<{ i: number }>('edge-pipe');
  const jobs = await Promise.all(Array.from({ length: 200 }, (_, i) => queue.add('p', { i })));
  const ids = new Set(jobs.map((j) => j.id));
  assertEq(ids.size, 200, 'all 200 ids unique');
  await waitFor(async () => (await queue.count()) === 200);
  queue.close();
});

test('edge: addBulk 1000 jobs in one frame', async () => {
  const queue = makeQueue<{ i: number }>('edge-bulk');
  const jobs = await queue.addBulk(
    Array.from({ length: 1000 }, (_, i) => ({ name: 'b', data: { i } }))
  );
  assertEq(jobs.length, 1000, '1000 ids returned');
  await waitFor(async () => (await queue.count()) === 1000);
  queue.close();
});

test('edge: multi-queue isolation on a shared connection', async () => {
  const a = makeQueue('edge-iso-a');
  const b = new Queue(qname('edge-iso-b'), { host: '127.0.0.1', port: getPort() });
  await a.add('only-a', { q: 'a' });
  await b.add('only-b', { q: 'b' });
  assertEq(await a.count(), 1, 'queue a sees 1');
  assertEq(await b.count(), 1, 'queue b sees 1');
  await a.obliterate();
  assertEq(await b.count(), 1, 'obliterating a leaves b intact');
  a.close();
  b.close();
});

test('edge: long-poll PULL on empty queue returns null after timeout', async () => {
  const queue = makeQueue('edge-poll');
  const started = Date.now();
  const response = await queue.connection.call(
    { cmd: 'PULL', queue: queue.name, owner: 'edge', timeout: 800 },
    10_000
  );
  assert(Date.now() - started >= 700, 'server held the long-poll');
  assertEq(response.job ?? null, null, 'no job on empty queue');
  queue.close();
});

test('edge: getJobs pagination is stable', async () => {
  const queue = makeQueue<{ i: number }>('edge-page');
  await queue.addBulk(Array.from({ length: 25 }, (_, i) => ({ name: 'p', data: { i } })));
  await waitFor(async () => (await queue.getJobs({ state: 'waiting' })).length === 25);
  const page1 = await queue.getJobs({ state: 'waiting', start: 0, end: 10 });
  const page2 = await queue.getJobs({ state: 'waiting', start: 10, end: 20 });
  const page3 = await queue.getJobs({ state: 'waiting', start: 20, end: 30 });
  assertEq(page1.length, 10, 'page1 = 10');
  assertEq(page2.length, 10, 'page2 = 10');
  assertEq(page3.length, 5, 'page3 = 5');
  const all = new Set([...page1, ...page2, ...page3].map((j) => j.id));
  assertEq(all.size, 25, 'pages do not overlap');
  queue.close();
});

test('edge: worker survives server crash + restart on the same port', async () => {
  // Dedicated server on a FIXED port so we can kill it and bring it back —
  // exactly what a production deploy/crash looks like to a connected client.
  const port = 21_000 + Math.floor(Math.random() * 5_000);
  const boot = () => spawnFixedServer(port);
  const first = await boot();
  const name = qname('edge-restart');
  const queue = new Queue<{ n: number }>(name, { host: '127.0.0.1', port });
  const seen: number[] = [];
  const { Worker } = await import('../dist/legacy.js');
  const worker = new Worker<{ n: number }, string>(
    name,
    (job) => {
      seen.push(job.data.n);
      return 'ok';
    },
    { host: '127.0.0.1', port, pollTimeoutMs: 300 }
  );
  worker.on('error', () => {}); // connection errors during the outage are expected
  await queue.add('before', { n: 1 });
  await waitFor(() => seen.includes(1));

  first.kill();
  await sleep(500);
  const second = await boot(); // same port: clients must reconnect lazily
  try {
    await waitFor(async () => {
      try {
        await queue.add('after', { n: 2 });
        return true;
      } catch {
        return false;
      }
    }, 20_000);
    await waitFor(() => seen.includes(2), 20_000);
    assert(seen.includes(1) && seen.includes(2), 'jobs before and after restart processed');
  } finally {
    await worker.close();
    queue.close();
    second.kill();
  }
});

test('edge: duplicate custom ids across addBulk are idempotent', async () => {
  const queue = makeQueue<{ i: number }>('edge-dup');
  const first = await queue.add('one', { i: 1 }, { jobId: 'edge-dup-key' });
  const again = await queue.add('one', { i: 2 }, { jobId: 'edge-dup-key' });
  assertEq(again.id, first.id, 're-add with same jobId returns the same job');
  assertEq(await queue.count(), 1, 'no duplicate enqueued');
  queue.close();
});

test('edge: results preserved per job under concurrency', async () => {
  const queue = makeQueue<{ i: number }>('edge-results');
  const ids: string[] = [];
  for (let i = 0; i < 20; i += 1) {
    ids.push((await queue.add('r', { i })).id);
  }
  const worker = makeWorker<{ i: number }, { double: number }>(
    queue.name,
    (job) => ({
      double: job.data.i * 2,
    }),
    { concurrency: 5 }
  );
  try {
    await waitFor(async () => (await queue.getJobCounts()).completed === 20, 20_000);
    for (let i = 0; i < 20; i += 1) {
      const result = (await queue.getResult(ids[i])) as { double: number };
      assertEq(result.double, i * 2, `result for job ${i} matches its input`);
    }
  } finally {
    await worker.close();
    queue.close();
  }
});
